components/support/nimbus-fml/src/parser.rs

```diff
@@ -2,12 +2,237 @@
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at http://mozilla.org/MPL/2.0/. */
 
-use std::fs::File;
+use std::{collections::HashMap, convert::TryFrom, path::Path};
 
-pub struct Parser {}
+use serde::{Deserialize, Serialize};
+
+use crate::{
+    error::FMLError,
+    intermediate_representation::{
+        EnumDef, FeatureDef, FeatureManifest, PropDef, TypeRef, VariantDef,
+    },
+};
+
+#[derive(Debug, Deserialize, Serialize)]
+pub(crate) struct EnumVariantBody {
+    description: String,
+}
+
+#[derive(Debug, Deserialize, Serialize)]
+pub(crate) struct EnumBody {
+    description: String,
+    variants: HashMap<String, EnumVariantBody>,
+}
+
+#[derive(Debug, Deserialize, Serialize)]
+pub(crate) struct VariableType(String);
+
+impl TryFrom<String> for TypeRef {
+    fn try_from(val: String) -> Result<TypeRef, FMLError> {
+        let (type_ref, type_name) = parse_typeref_string(val)?;
+
+        return match type_ref.as_str() {
+            "String" => Ok(TypeRef::String),
+            "Int" => Ok(TypeRef::Int),
+            "Boolean" => Ok(TypeRef::Boolean),
+            "BundleText" => Ok(TypeRef::BundleText(type_name.unwrap())),
+            "BundleImage" => Ok(TypeRef::BundleImage(type_name.unwrap())),
+            "Enum" => Ok(TypeRef::Enum(type_name.unwrap())),
+            "Object" => Ok(TypeRef::Object(type_name.unwrap())),
+            "List" => Ok(TypeRef::List(Box::new(TypeRef::try_from(
+                type_name.unwrap(),
+            )?))),
+            "Option" => Ok(TypeRef::Option(Box::new(TypeRef::try_from(
+                type_name.unwrap(),
+            )?))),
+            "Map" => {
+                // Maps take a little extra massaging to get the key and value types
```

Comment:
Yeah, this looks narlier than it needs to be :(

Let's get rid of `StringMap` and rename `EnumMap` to `Map`. What was I thinking.